### ISSUE TITLE
fewest-build-containers: tasks in aggregate land on the same worker

### ISSUE THREAD

**Issue by wagdav** [concourse/concourse#3301]
## Bug Report

The "fewest-build-container" placement strategy always puts tasks within an aggregate block to the same worker.

## Steps to Reproduce

* enable the fewest-build-container strategy
* create a pipeline with an aggregate block

```
- name: job with aggregate
  plan:
  - aggregate:
    - task: build-on-ubuntu-16.04
    - task: build-on-ubuntu-18.04
```

Schedule some builds and observe the output of `fly containers`

## Expected Results

Tasks within an aggregate respect the fewest-build-container strategy. Sometimes they _may_ land on the same worker.

## Actual Results

Tasks within an aggregate _always_ land on the same worker.

## Additional Context

A sample output of `fly containers`:

```
63aebe28-b48c-4ac1-6d33-ba1a84300cac  wrk-lin-0-07b4efc347e46c064      pipeline-master                                   build                      81    84161  task   build-ubuntu-18.04                        n/a
71e38b7e-aaf0-42db-462e-7d2a361ec90a  wrk-lin-0-07b4efc347e46c064      pipeline-master                                   build                      81    84161  task   build-ubuntu-16.04                        n/a
[...]
ddd3b498-b3f8-44ce-4735-246937b4b884  wrk-lin-3-0544d8e8cdbc32852      pipeline-master                                   conan-pkg                  15    84179  task   build-ubuntu-16.04                        n/a
e720603b-537c-4c49-4be1-cdeb37402155  wrk-lin-3-0544d8e8cdbc32852      pipeline-master                                   conan-pkg                  15    84179  task   build-ubuntu-18.04                        n/a
```

## Version Info

* Concourse version: 5.0 RC 74
* Deployment type (BOSH/Docker/binary): binary
* Infrastructure/IaaS: AWS
* Browser (if applicable):
* Did this used to work? no.

**Comment by kcmannem** [comment]
Thanks for pointing this out! It makes sense that we see this because the `fewest-build-containers` strategy depends upon the number of build containers on a worker. https://github.com/concourse/concourse/blob/master/atc/worker/placement.go#L75-L81

Since the aggregate steps are scheduled in parallel, the number of build containers on each worker is the same when scheduling all of them, and so the placement strategy selects the same worker for each one.
https://github.com/concourse/concourse/blob/master/atc/exec/aggregate.go#L22-L27

To fix this, the strategy would need to be aware of other containers being created on the same worker, which is currently not possible. We created another issue, #3304, to capture this work as well.

**Comment by wagdav** [comment]
@kcmannem Thanks for the quick response! I see your reasoning.  I'm happy that you're considering this for the next milestone.

**Comment by wagdav** [comment]
@kcmannem Our team greatly appreciates that you point to specific parts of the code.  We learn a lot from your comments!

**Comment by ddadlani** [comment]
As explained above, the placement strategy has no knowledge of the type of step for which a container is being created, so it is hard to account for this case where multiple containers are created in parallel.

After some investigation, we are going to try pulling the container placement logic up a level, such that each type of 'meta-step' (aggregate, ensure, on_failure etc) can place containers differently.

Our proposed approach is to separate container creation and step running logic. Currently, each of `get`, `put` and `task` step's `Run()` method creates the container and volumes and then runs the step. If we break this up and add a `Create` or `Setup` method to the `Step` interface: 

https://github.com/concourse/concourse/blob/b80b93953505358de8b271b2fdbf94a867fd5934/atc/exec/step.go#L16-L29

We can then call the `Create/Setup` method first in all the `meta-step` types separately for each child step and then run the step.

The disadvantage of this approach is that the `Create/Setup` method will need to be no-op for all the `meta-step` types but we don't have any ideas on how to mitigate that at the moment. All input is welcome.

**Comment by ddadlani** [comment]
Based on a conversation with @pivotal-jwinters and @topherbullock, @kcmannem and I think that this is a good time to invert control of step running as well. We will look into adding a `StepRunner` interface which orchestrates the required containers and runs each step in its container. 

This is different from the current implementation in that `step`s will no longer `Run()` themselves. It will allow for a cleaner separation of container creation and step running.

**Comment by vito** [comment]
Hmm this feels like a more fundamental problem, and I don't think adding logic to different step types is going to help much. This could also have happened if you have 10 jobs that all fired concurrently without any use of `aggregate`.

As long as the selection of the worker and the 'allocation' of the container isn't done in one atomic step, I think this problem is going to still exist at some level.

The step interface right now is one of the simpler and better-documented parts of the Concourse codebase, so growing the interface and adding no-op methods and having more work to do in the 'meta' steps kind of gives me pause.

**Comment by kcmannem** [comment]
We've pulled up initialization portion(essentially placing a row in the container table) from `worker.FindOrCreateContainer` into `pool.FindOrCreatingContainer`. This allows us to place a "worker selection lock" around choosing a worker and updating db state. This way any concurrent build triggers and aggregate steps choose a worker based on a new state in the db. This shouldn't impact performance very much as creating physical containers is not serialized.

**Comment by topherbullock** [comment]
TODO: 
- [ ] Create an integration test suite for the runtime + container scheduling to reduce the overhead on topgun, and increase our confidence in the selection of workers for task containers using this strategy

**Comment by ddadlani** [comment]
FYI we theorize that this happened to Strabo on wings too (https://wings.pivotal.io/teams/cf-tools/pipelines/strabo/jobs/generate-pivotal-feeds/builds/145). The error is `iptables: create-instance-chains: iptables: Memory allocation problem`. Around the time this error occurred, we saw a spike in containers in one of the strabo workers, which went up to around 150 containers.

We think this might be happening because Strabo kicks off a bunch of get containers at the same time, and because of this bug, they all get scheduled on the same worker which causes the above error.

We saw this error earlier in #3127, but that was caused by check containers and that has already been fixed.

**Comment by ddadlani** [comment]
The only way to fix this issue would be to lock worker allocation and `db.CreatingContainer` creation. After the introduction of https://github.com/concourse/concourse/pull/3373, worker allocation (`pool.FindOrChooseWorker`) and CreatingContainer creation do not always occur at the same point in the code, making this a hard problem to solve.

This is now blocked on #3424, because the introduction of ephemeral containers reduces the number of places where this needs to happen.

**Comment by vito** [comment]
There was a race condition introduced by the fix where the container owner can't be found in some cases. This is still pending investigation --by @ddadlani

**Comment by ddadlani** [comment]
There is an unintended side effect to container creation where we don't find an existing check container if its resource config check session is about to expire, but we reuse the same expiring check session as an owner to create a new check container (which then expires shortly after). This race always existed but is now exposed as an error in our work around adding a lock in FindOrCreateContainer. Because `FindOrCreateCheckSession` has been removed from the resource scan flow, there is no easy way to ensure that when a check session is expiring, a new one is created as an owner for the new check container. 

Because this entire flow will be made redundant by #3079 , we are going to block this on #3079 and give that a shot before the `artifacts` flow. If that proves difficult, we will find another workaround to fix this issue

**Comment by ddadlani** [comment]
The fix we made for this issue broke `fly containers` because it left around extra `db.CreatingContainers` in the database. We attempted to fix db container creation but because of the way the worker is passed into the [`Fetcher`](https://github.com/concourse/concourse/blob/a1b5da0fe1c1346a3ef3eac19a92bde1755e240f/atc/exec/get_step.go#L196), this becomes difficult. We need a worker for the [`fetchSource.Find()`](https://github.com/concourse/concourse/blob/a1b5da0fe1c1346a3ef3eac19a92bde1755e240f/atc/resource/resource_instance_fetch_source.go#L87), but we may not need a container. This means that we cannot create a creating container during worker selection.

I am reopening this issue so that we can find another solution to the fewest-build-containers aggregate step problem.
